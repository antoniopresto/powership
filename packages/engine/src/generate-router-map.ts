import { glob } from 'zx';
import { RouteUtils } from '@powership/utils';

export async function generateRouterMap(sourceDir: string) {
  const found = await glob(['{pages,api}/**/*.{ts,tsx}'], {
    cwd: sourceDir,
  });

  const sorted = RouteUtils.sortRoutes(found);

  return sorted.map((filePath) => {
    const routePath = filePath.replace(/\[([^\]]+)\](\.tsx?)?/g, ':$1');

    const routeMatcher = RouteUtils.createRouteMatcher(routePath);

    return {
      filePath,
      path: routePath,
      match: routeMatcher.match.bind(routeMatcher),
      stringify: routeMatcher.stringify.bind(routeMatcher),
    };
  });
}
