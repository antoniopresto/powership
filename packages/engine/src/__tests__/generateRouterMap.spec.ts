import { generateRouterMap } from '../generate-router-map';
import * as path from 'node:path';
import * as process from 'node:process';

describe('generateRouterMap', () => {
  // afterEach();

  test('basic test', async () => {
    const sut = await generateRouterMap(
      path.resolve(process.cwd(), 'src/__tests__/app0/src')
    );

    expect(sut).toMatchSnapshot();
  });
});
