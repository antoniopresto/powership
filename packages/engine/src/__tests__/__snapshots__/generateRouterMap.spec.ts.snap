// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRouterMap basic test 1`] = `
[
  {
    "filePath": "pages/account.tsx",
    "match": [Function],
    "path": "pages/account.tsx",
    "stringify": [Function],
  },
  {
    "filePath": "pages/home.tsx",
    "match": [Function],
    "path": "pages/home.tsx",
    "stringify": [Function],
  },
  {
    "filePath": "api/product/[id].tsx",
    "match": [Function],
    "path": "api/product/:id",
    "stringify": [Function],
  },
  {
    "filePath": "pages/[category]/[slug].tsx",
    "match": [Function],
    "path": "pages/:category/:slug",
    "stringify": [Function],
  },
]
`;
